src/v/cluster/non_replicable_partition.cc

```diff
@@ -0,0 +1,111 @@
+ * the Business Source License, use of this software will be governed
+ * by the Apache License, Version 2.0
+ */
+
+#include "cluster/non_replicable_partition.h"
+
+#include "storage/log.h"
+
+namespace {
+
+/// Ensures that operations on a model::record_batch_reader occur within the
+/// context of a gate for safe shutdown. If operations are observed while the
+/// gate is closed the caller will observe the ss::gate_closed_exception
+class gated_shutdown_reader final : public model::record_batch_reader::impl {
+public:
+    gated_shutdown_reader(
+      ss::gate& g,
+      std::unique_ptr<model::record_batch_reader::impl> impl) noexcept
+      : _gate(g)
+      , _impl(std::move(impl)) {}
+
+    bool is_end_of_stream() const final {
+        return _gate.is_closed() || _impl->is_end_of_stream();
+    }
+
+    ss::future<model::record_batch_reader::storage_t>
+    do_load_slice(model::timeout_clock::time_point t) final {
+        auto holder = _gate.hold();
+        return _impl->do_load_slice(t).finally([holder] {});
+    }
+
+    void print(std::ostream& os) final { os << "gated_shutdown_reader"; }
+
+private:
+    ss::gate& _gate;
+    std::unique_ptr<model::record_batch_reader::impl> _impl;
+};
+
+/// Not neccessary to hold gate open until all processing completes, as that can
+/// be done by invoking the write within the context of the \ref
+/// non_replicable_partition gate. This class just makes the operation
+/// cancellable
+class gated_shutdown_writer {
+public:
+    gated_shutdown_writer(
+      ss::gate& g, storage::log_appender&& appender) noexcept
+      : _gate(g)
+      , _appender(std::move(appender)) {}
+
+    ss::future<ss::stop_iteration> operator()(model::record_batch& rb) {
+        if (_gate.is_closed()) {
+            return ss::make_ready_future<ss::stop_iteration>(
+              ss::stop_iteration::yes);
+        }
+        return _appender.operator()(rb);
+    }
+
+    ss::future<storage::append_result> end_of_stream() {
+        return _appender.end_of_stream();
+    }
+
+private:
+    ss::gate& _gate;
+    storage::log_appender _appender;
+};
+
+} // namespace
+
+namespace cluster {
+non_replicable_partition::non_replicable_partition(
+  storage::log log, ss::lw_shared_ptr<partition> source) noexcept
+  : _log(log)
+  , _source(source) {}
+
+ss::future<model::record_batch_reader>
+non_replicable_partition::make_reader(storage::log_reader_config config) {
+    return ss::with_gate(_gate, [this, config = std::move(config)] {
+        return _log.make_reader(config).then(
+          [this](model::record_batch_reader rdr) {
+              return model::make_record_batch_reader<gated_shutdown_reader>(
+                _gate, std::move(rdr).release());
+          });
+    });
+}
+
+ss::future<storage::append_result>
+non_replicable_partition::write(model::record_batch_reader&& rbr) {
+    return ss::with_gate(_gate, [this, rbr = std::move(rbr)]() mutable {
+        const storage::log_append_config write_cfg{
```

Comment:
Maybe the API to this method should accept a `storage::log_append_config` ?

Reply:
Or better yet, replace the method with `make_appender` which returns the consumer impl above?